src/category_theory/limits/shapes/cartesian_closed.lean

```diff
@@ -0,0 +1,387 @@
+adjunction.hom_equiv_unit _
+
+lemma uncurry_id_eq_ev (A X : C) [exponentiable A] : uncurry (𝟙 (A ⟹ X)) = ev A X :=
+by rw [uncurry_eq, prod_map_id_id, id_comp]
+
+lemma curry_id_eq_coev (A X : C) [exponentiable A] : curry (𝟙 _) = coev A X :=
+by { rw [curry_eq, post, functor.map_id], apply comp_id }
+
+lemma curry_injective : function.injective (curry : (A ⨯ Y ⟶ X) → (Y ⟶ A ⟹ X)) :=
+(exponentiable.is_adj.adj.hom_equiv _ _).injective
+
+lemma uncurry_injective : function.injective (uncurry : (Y ⟶ A ⟹ X) → (A ⨯ Y ⟶ X)) :=
+(exponentiable.is_adj.adj.hom_equiv _ _).symm.injective
+
+section terminal
+
+/--
+The terminal object is always exponentiable.
+This isn't an instance because most of the time we'll prove cartesian closed for all objects
+at once, rather than just for this one.
+-/
+def terminal_exponentiable : exponentiable ⊤_C :=
+{ is_adj :=
+  { right := 𝟭 C,
+    adj := adjunction.mk_of_hom_equiv
+    { hom_equiv := λ X _, have unitor : _, from prod.left_unitor X,
+        ⟨λ a, unitor.inv ≫ a, λ a, unitor.hom ≫ a, by tidy, by tidy⟩ } } }
+
+/-- The typeclass argument is explicit: any instance can be used, not just the above. -/
+def exp_terminal_iso [exponentiable ⊤_C] : (⊤_C ⟹ X) ≅ X :=
+yoneda.ext (⊤_ C ⟹ X) X
+  (λ Y f, (prod.left_unitor Y).inv ≫ uncurry f)
+  (λ Y f, curry ((prod.left_unitor Y).hom ≫ f))
+  (λ Z g, by rw [curry_eq_iff, iso.hom_inv_id_assoc] )
+  (λ Z g, by simp)
+  (λ Z W f g, by rw [uncurry_natural_left, prod_left_unitor_inv_naturality_assoc f] )
+
+/-- The internal element which points at the given morphism. -/
+@[reducible]
+def point_at_hom (f : A ⟶ Y) : ⊤_C ⟶ (A ⟹ Y) :=
+curry (limits.prod.fst ≫ f)
+
+end terminal
+
+section pre
+
+variables [has_finite_products.{v} C] {B}
+
+/-- Pre-compose an internal hom with an external hom. -/
+def pre (X : C) (f : B ⟶ A) [exponentiable A] [exponentiable B] :  (A⟹X) ⟶ B⟹X :=
+curry (limits.prod.map f (𝟙 _) ≫ ev A X)
+
+lemma pre_id (A X : C) [exponentiable A] : pre X (𝟙 A) = 𝟙 (A⟹X) :=
+by { rw [pre, prod_map_id_id, id_comp, ← uncurry_id_eq_ev], simp }
+
+-- There's probably a better proof of this somehow
+/-- Precomposition is contrafunctorial. -/
+lemma pre_map [exponentiable A] [exponentiable B] {D : C} [exponentiable D] (f : A ⟶ B) (g : B ⟶ D) :
+  pre X (f ≫ g) = pre X g ≫ pre X f :=
+begin
+  rw [pre, curry_eq_iff, pre, pre, uncurry_natural_left, uncurry_curry, prod_map_map_assoc,
+      prod_map_comp_id, assoc, ← uncurry_id_eq_ev, ← uncurry_id_eq_ev, ← uncurry_natural_left,
+      curry_natural_right, comp_id, uncurry_natural_right, uncurry_curry],
+end
+
+end pre
+
+/-- The precomposition functor. -/
+@[simps]
+def pre.functor [is_cartesian_closed C] (X : C) : Cᵒᵖ ⥤ C :=
+{ obj := λ A, (A.unop) ⟹ X,
+  map := λ A B f, pre X f.unop,
+  map_id' := λ B, pre_id B.unop X,
+  map_comp' := λ P Q R f g, pre_map g.unop f.unop }
+
+lemma exp_natural [is_cartesian_closed C] {A B : C} {X Y : Cᵒᵖ} (f : A ⟶ B) (g : X ⟶ Y) :
```

Comment:
As discussed above, we should try to be consistent about using `nat`/`natural`/`naturality`.

Reply:
Yup, I changed this name because I think it wasn't the correct name anyway